# Table of Contents
1. Importing Libraries
2. Importing Files
3. Data Consistency Checks
4. Answers of Exercise 4.5
5. Exporting Checked Files

## 01 Importing Libraries

In [22]:
import pandas as pd
import numpy as np
import os

## 02 Importing files 

In [23]:
path = r'/Users/niravbariya/Documents/Data Science/CareerFoundry/Data Analytics Immersion/Achievement 4/04-04-2024 Instacart Basket Analysis'

In [24]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [25]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [26]:
df_prods.shape

(49693, 5)

## 03 Data Consistency Checks

### Mixed Data types

In [27]:
df_test = pd.DataFrame()

In [28]:
df_test['mix'] = ['a', 'b', 1, True]

### Missing Values

In [29]:
#Finding missing values in all columns
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [11]:
# Looking at the missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [12]:
df_nan


product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [15]:
# Dimension of df_prods
df_prods.shape

(49693, 5)

In [16]:
# Creating a clean dataframe without missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [17]:
# Dimensions of clean dataframe
df_prods_clean.shape

(49677, 5)

### Handling Duplicates

In [18]:
# Extracting duplicate values 
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [19]:
# Checking the dimensions
df_prods_clean.shape

(49677, 5)

In [31]:
# How many duplicates in prods?
df_prods_clean[df_prods_clean.duplicated()].shape

In [20]:
# Creating a dataframe without duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [21]:
# Dimensions of dataframe without duplicates or null values
df_prods_clean_no_dups.shape

(49672, 5)

In [63]:
# Exporting checked product dataframe to a csv file
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data','products_checked.csv'), index = False)

## 04 Exercise 4.5 Answers

In [16]:
#reading the file orders_wrangled.csv
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [17]:
#Describing the data 
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

Minimum of days_since_prior_order can not be 0. For the first order it will be NaN. After that, it will always be some positive integer. This should be investigated further.

### Checking for Mixed type in orders

In [18]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

/var/folders/xd/gd6kszh506xbhlz9mv23vpxh0000gn/T/ipykernel_79590/1946670780.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/xd/gd6kszh506xbhlz9mv23vpxh0000gn/T/ipykernel_79590/1946670780.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/xd/gd6kszh506xbhlz9mv23vpxh0000gn/T/ipykernel_79590/1946670780.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/xd/gd6kszh506xbhlz9mv23vpxh0000gn/T/ipykernel_79590/1946670780.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].ilo

There are no columns with mixed data types.

### Checking for Missing Values in orders

In [19]:
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

The missing values are in days_since_prior_order. We can expect NaN values in this column for the first time customers as they will
not have any previous order. Total user_ids are 206209, that is same as the missing values found in days_since_prior_order column.
This makes it clear that every user_id will have one NaN value in this column conforming our guess that first orders will have NaN.

### Addressing the missing values
We will not remove the duplicate values as it contains useful data about orders made in other columns. We can not replace NULL values with 0 as that will have impact while doing any kind of aggregation on the column. I think we can simply keep this NaN as flag as it represents 
first orders.

### Checking for duplicates

In [20]:
df_ords[df_ords.duplicated()]

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [21]:
df_ords.shape

(3421083, 6)

There are no full duplicates in orders dataframe. There are rows with same user_ids but that will not be counted as duplicate entry since the 
order_id is the unique identifier.

### Addressing Duplicates

Since there are no full duplicate rows and repeated user_ids is the nature of the dataframe, there is no need of taking any action except for keeping a record of the observation.

## 05. Exporting Checked files

In [64]:
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'), index = False)

In [65]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data','products_checked.csv'), index = False)